File: EPI/9.9/swift/9.9_test.swift
import Darwin

func == <T: Equatable>(lhs: [[T]], rhs: [[T]]) -> Bool {
    guard lhs.count == rhs.count else {
        return false
    }
    for (index, values) in lhs.enumerated() {
        guard rhs[index] == values else {
            return false
        }
    }
    return true
}

// Tree:
//           5
//        /     \
//     42         36
//    /  \       /
//   4    12   15
//  / \       /  \
// 2   9     3    101
//                   \
//                    102

let root = Node(
    value: 5,
    left: Node(
        value: 42,
        left: Node(
            value: 4,
            left: Node(value: 2),
            right: Node(value: 9)
        ),
        right: Node(value: 12)
    ),
    right: Node(
        value: 36,
        left: Node(
            value: 15,
            left: Node(value: 3),
            right: Node(
                value: 101,
                right: Node(value: 102)
            )
        )
    )
)

let expected = [
    [5],
    [42, 36],
    [4, 12, 15],
    [2, 9, 3, 101],
    [102],
]

let actual = root.valuesByIncreasingDepth()

guard actual == expected else {
    print("Expected values in order of increasing depth to be \(expected), but were \(actual)")
    exit(1)
}

File: EPI/Stacks and Queues/Compute binary tree nodes in order of increasing depth/typescript/9.9_test.ts
import {Node, getNodesByDepth} from './9.9';

declare function require(name: string): any;
const assert = require('assert');

// Tree:
//           A
//         /   \
//       B       C
//      / \     /
//     D   E   F
//    / \     / \
//   G   H   I   J
//                \
//                 K

const [A, B, C, D, E, F, G, H, I, J, K]: Node[] = Array.from({length: 11}, () => ({}));

A.left = B;
A.right = C;

B.left = D;
B.right = E;

D.left = G;
D.right = H;

C.left = F;

F.left = I;
F.right = J;

J.right = K;

const actual = getNodesByDepth(A);

const expected = [
    [A],
    [B, C],
    [D, E, F],
    [G, H, I, J],
    [K]
];

assert.deepStrictEqual(actual, expected);
